opentelemetry/tracing_interceptor.go

```diff
@@ -0,0 +1,202 @@
+	// global call.
+	TextMapPropagator propagation.TextMapPropagator
+
+	// SpanContextKey is the context key used for internal span tracking (not to
+	// be confused with the context key OpenTelemetry uses internally). If not
+	// set, this defaults to an internal key (recommended).
+	SpanContextKey interface{}
+
+	// HeaderKey is the Temporal header field key used to serialize spans. If
+	// empty, this defaults to the one used by all SDKs (recommended).
+	HeaderKey string
+
+	// SpanStarter is a callback to create spans. If not set, this creates normal
+	// OpenTelemetry spans calling Tracer.Start.
+	SpanStarter func(ctx context.Context, t trace.Tracer, spanName string, opts ...trace.SpanStartOption) trace.Span
+}
+
+type spanContextKey struct{}
+
+const defaultHeaderKey = "_tracer-data"
+
+type tracer struct{ options *TracerOptions }
+
+// NewTracer creates a tracer with the given options. Most callers should use
+// NewTracingInterceptor instead.
+func NewTracer(options TracerOptions) (interceptor.Tracer, error) {
+	if options.Tracer == nil {
+		options.Tracer = otel.GetTracerProvider().Tracer("temporal-sdk-go")
+	}
+	if options.TextMapPropagator == nil {
+		options.TextMapPropagator = DefaultTextMapPropagator
+	}
+	if options.SpanContextKey == nil {
+		options.SpanContextKey = spanContextKey{}
+	}
+	if options.HeaderKey == "" {
+		options.HeaderKey = defaultHeaderKey
+	}
+	if options.SpanStarter == nil {
+		options.SpanStarter = func(
+			ctx context.Context,
+			t trace.Tracer,
+			spanName string,
+			opts ...trace.SpanStartOption,
+		) trace.Span {
+			_, span := t.Start(ctx, spanName, opts...)
+			return span
+		}
+	}
+	return &tracer{&options}, nil
+}
+
+// NewTracingInterceptor creates an interceptor for setting on client options
+// that implements OpenTelemetry tracing for workflows.
+func NewTracingInterceptor(options TracerOptions) (interceptor.Interceptor, error) {
+	t, err := NewTracer(options)
+	if err != nil {
+		return nil, err
+	}
+	return interceptor.NewTracingInterceptor(t), nil
+}
+
+func (t *tracer) Options() interceptor.TracerOptions {
+	return interceptor.TracerOptions{
+		SpanContextKey: t.options.SpanContextKey,
+		HeaderKey:      t.options.HeaderKey,
+	}
+}
+
+func (t *tracer) UnmarshalSpan(m map[string]string) (interceptor.TracerSpanRef, error) {
+	ctx := trace.SpanContextFromContext(t.options.TextMapPropagator.Extract(context.Background(), textMapCarrier(m)))
+	if !ctx.IsValid() {
+		return nil, nil
+	}
+	return &tracerSpanRef{SpanContext: ctx}, nil
+}
+
+func (t *tracer) MarshalSpan(span interceptor.TracerSpan) (map[string]string, error) {
+	data := textMapCarrier{}
+	t.options.TextMapPropagator.Inject(trace.ContextWithSpan(context.Background(), span.(*tracerSpan).Span), data)
+	return map[string]string(data), nil
+}
+
+func (t *tracer) SpanFromContext(ctx context.Context) interceptor.TracerSpan {
+	span := trace.SpanFromContext(ctx)
+	if !span.SpanContext().IsValid() {
+		return nil
```

Comment:
Add some debug log?

Reply:
There can be plenty of normal cases where a context doesn't have a span yet. From `interceptor.Tracer.SpanFromContext` docs:

> SpanFromContext returns the span from the general Go context or nil if not present.

Absence of a span is considered normal.